doc/safegcd_implementation.md

```diff
@@ -126,6 +126,27 @@ The approach here avoids the need for 2x2 matrix multiplications of various size
 be faster at the level of optimization we're able to do in C.
 
 
+## 2b. From GCDs to Jacobi symbol
+
+We can also use a similar approach to calculate Jacobi symbol (x | M) by keeping track of an extra variable j, for which at every step (x | M) = j * (g | f). As we update f and g, we make corresponding updates to j using [properties of the Jacobi symbol](https://en.wikipedia.org/wiki/Jacobi_symbol#Properties). In particular, we update j whenever we divide g by 2 or swap f and g; these updates depend only on the values of f and g modulo 4 or 8, and can thus be applied very quickly. Overall, this calculation is slightly simpler than the one for modular inverse because we no longer need to keep track of d, e.
+
+However, one difficulty of this approach is that the Jacobi symbol (a | n) is only defined for positive odd integers n, whereas in the original safegcd algorithm, f, g can take negative values. We resolve this by making the following change to the gcd algorithm:
+
+```python
+        # Before
+        if delta > 0 and g & 1:
+            delta, f, g = 1 - delta, g, (g - f) // 2
+
+        # After
+        if delta > 0 and g & 1:
+            delta, f, g = 1 - delta, g, (g + f) // 2
+```
+
+The algorithm is still correct, since the changed divstep (which we will call a "positive division step", or "posdivstep") preserves gcd(f, g). However, there's no proof that the changed algorithm will converge. The justification for posdivsteps is completely empirical: in practice, all out of 2.5 billion random inputs (with the field size modulus) converged to f=g=gcd(x, M). We account for the possibility of nonconvergence by only performing a bounded number of posdivsteps, and then falling back to square-root based Jacobi calculation if a solution has not yet been found. Finally, since f, g will stop changing once f=g=gcd(f, g) we need to update the termination condition from g=0 to f=1 (we require inputs to satisfy gcd(x, M) = 1).
```

Comment:
The name "posdivsteps" actually comes from the safegcd paper (section 8.4). It may be useful to refer to it (though it calls it non-functioning because it fails to send g to 0, and apparently misses that it instead converges to f=g).

I don't think we should mention the full analysis here; maybe just "in practice, it appears that the vast majority of inputs converge in a number of steps proportional to their logarithm"?

Reply:
Good idea, done!